solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs

```diff
@@ -1,26 +1,81 @@
+use anchor_lang::prelude::{CpiContext, Pubkey};
 use anchor_lang::solana_program::msg;
+use anchor_spl::token::{spl_token, Burn, MintTo, Transfer};
 use ibc::applications::transfer::context::{
     TokenTransferExecutionContext, TokenTransferValidationContext,
 };
 use ibc::applications::transfer::error::TokenTransferError;
 use ibc::applications::transfer::PrefixedCoin;
 use ibc::core::ics24_host::identifier::{ChannelId, PortId};
 use ibc::Signer;
+use uint::FromDecStrErr;
 
 // use crate::module_holder::IbcStorage<'_,'_>;
-use crate::IbcStorage;
+use crate::{id, IbcStorage, MINT_ESCROW_SEED};
 
-impl TokenTransferExecutionContext for IbcStorage<'_, '_> {
+impl TokenTransferExecutionContext for IbcStorage<'_, '_, '_, '_> {
     fn send_coins_execute(
         &mut self,
-        _from: &Self::AccountId,
-        _to: &Self::AccountId,
-        _amt: &PrefixedCoin,
+        from: &Self::AccountId,
+        to: &Self::AccountId,
+        amt: &PrefixedCoin,
     ) -> Result<(), TokenTransferError> {
-        //let sender_id = from.to_string();
-        //let receiver_id = to.to_string();
-        //let base_denom = amt.denom.base_denom.to_string();
-        todo!()
+        msg!(
+            "Sending coins from account {} to account {}, trace path {}, base \
+             denom {}",
+            from,
+            to,
+            amt.denom.trace_path,
+            amt.denom.base_denom
+        );
+        let sender_id = from.to_string();
+        let receiver_id = to.to_string();
+        let base_denom = amt.denom.base_denom.to_string();
+        let amount = amt.amount;
+        // Since amount is u256 which is array of u64, so if the amount is above u64 max, it means that the amount value at index 0 is max.
+        if amount[0] == u64::MAX {
+            return Err(TokenTransferError::InvalidAmount(
+                FromDecStrErr::InvalidLength,
+            ));
+        }
+        let (_token_mint_key, bump) =
+            Pubkey::find_program_address(&[base_denom.as_ref()], &id());
+        let store = self.0.borrow();
+        let sender = store
+            .accounts
+            .iter()
+            .find(|account| account.key.to_string() == sender_id)
+            .ok_or(TokenTransferError::ParseAccountFailure)?;
```

Comment:
This needs to be extracted to a function. Possibly method on IbcStorage

Reply:
Sure got it